### PR TITLE
chore: refactor runner and run.service

### PR DESCRIPTION
**Pull request by TBonnin**
## Describe your changes

First I would like to apologize for the size of the PR. 

The goal of this big PR is to make the communication between `jobs` and `runner` asynchronous, instead of relying on a long running request that can fail and depends on `jobs` in-memory state.

The cancellation/aborting mechanism had to be refactored also since it was also relying on an in-memory state in `jobs`

The `run.service` logic has been split into one file per script type. It is not the most DRY solution but it avoids the current situation where it is very hard to follow what part of run.service applies to which script type. They all differ slightly from each other and unless we have a good abstraction around them it is better imho to have a bit of duplication. Let me know what you think.

Notes:
- server/runner/orchestrator must be deployed first before jobs

## Issue ticket number and link
https://linear.app/nango/issue/NAN-1421/refactor-runservice

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/cli/lib/services/dryrun.service.ts [hunk 1]
```diff
@@ -499,12 +504,7 @@ export class DryRunService implements ScriptExecutorInterface {
                     return { success: false, error: err, response: null };
                 }
 
-                let errorType = 'sync_script_failure';
-                if (isWebhook) {
```
Comment (bodinsamuel) [inline packages/cli/lib/services/dryrun.service.ts:505]: Webhooks can't be dryrun?

File: packages/shared/lib/sdk/sync.integration.test.ts [hunk 2]
```diff
@@ -37,10 +37,13 @@ describe('Connection service integration tests', () => {
             }
 
             const nangoProps: NangoProps = {
+                scriptType: 'sync',
                 host: 'http://localhost:3003',
-                accountId: environment.account_id,
+                teamId: environment.account_id,
+                teamName: 'team',
```
Comment (bodinsamuel) [inline packages/shared/lib/sdk/sync.integration.test.ts]: Could have been great to pass `team: Pick<id, name>` instead

File: packages/jobs/lib/routes/tasks/putTask.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    Error: ApiError<'put_task_failed'>;
+    Success: never;
+}>;
+
+const jsonLiteralSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
+export const jsonSchema: z.ZodType<JsonValue> = z.lazy(() => z.union([jsonLiteralSchema, z.array(jsonSchema), z.record(jsonSchema)]));
+
+const validate = validateRequest<PutTask>({
+    parseBody: (data) =>
+        z
```
Comment (bodinsamuel) [inline packages/jobs/lib/routes/tasks/putTask.ts:88]: I'm wondering what's the overhead of re-creating this zod object all the time

File: packages/webhooks/lib/sync.ts [hunk 4]
```diff
@@ -80,7 +80,7 @@ export const sendSync = async ({
                 deleted: 0
             },
             modifiedAfter: dayjs(now).toDate().toISOString(),
-            queryTimeStamp: operation !== 'INITIAL' ? (now as unknown as string) : null
+            queryTimeStamp: now as unknown as string // Deprecated
```
Comment (khaliqgant) [inline packages/webhooks/lib/sync.ts:83]: does this need to change? Can you go through why this is needed?

Comment (TBonnin) [reply packages/webhooks/lib/sync.ts:83]: no sync is ever set as `INITIAL` anymore so this condition is always true and therefore unnecessary

File: packages/jobs/lib/scripts/operations/utils/getRunner.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import type { Runner } from '../../../runner/runner.js';
+import { getOrStartRunner, getRunnerId } from '../../../runner/runner.js';
+
+export async function getRunner(teamId: number): Promise<Result<Runner>> {
+    // a runner per account in prod only
+    const runnerId = isProd ? getRunnerId(`${teamId}`) : getRunnerId('default');
+    // fallback to default runner if account runner isn't ready yet
+    const runner = await getOrStartRunner(runnerId).catch(() => getOrStartRunner(getRunnerId('default')));
+    return Ok(runner);
+}
```
Comment (TBonnin) [inline packages/jobs/lib/scripts/operations/utils/getRunner.ts]: It is a util function because it is called by both `start` and `abort`

File: packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+const validate = validateRequest<PostHeartbeat>({
+    parseParams: (data) => z.object({ taskId: z.string().uuid() }).strict().parse(data)
+});
+
+const handler = async (req: EndpointRequest<PostHeartbeat>, res: EndpointResponse<PostHeartbeat>) => {
+    const { taskId } = req.params;
+    const heartbeat = await orchestratorClient.heartbeat({ taskId });
+    if (heartbeat.isErr()) {
+        res.status(400).json({ error: { code: 'heartbeat_failed', message: `heartbeat failed: taskId=${taskId} error='${heartbeat.error}'` } });
```
Comment (bodinsamuel) [inline packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts]: Note the json for error contains a payload/errors properties that you can use

File: packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    }
```
Comment (bodinsamuel) [inline packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts:31]: Missing return

File: packages/jobs/lib/execution/operations/output.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        case 'webhook':
+            await handleWebhookOutput({ nangoProps });
+            break;
+        case 'post-connection-script':
+            await handlePostConnectionOutput({ nangoProps });
+            break;
+    }
+    const setSuccess = await orchestratorClient.succeed({ taskId, output: output });
+    if (setSuccess.isErr()) {
+        handlePayloadTooBigError({ taskId, error: setSuccess.error, nangoProps });
```
Comment (bodinsamuel) [inline packages/jobs/lib/execution/operations/output.ts]:
```suggestion
        await handlePayloadTooBigError({ taskId, error: setSuccess.error, nangoProps });
```

File: packages/jobs/lib/routes/tasks/putTask.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    if (!nangoProps) {
+        res.status(400).json({ error: { code: 'put_task_failed', message: 'missing nangoProps' } });
+        return;
+    }
+    if (error) {
+        await handleError({ taskId, nangoProps, error });
+    } else {
+        await handleOutput({ taskId, nangoProps, output: output });
+    }
+    res.status(201).send();
```
Comment (bodinsamuel) [inline packages/jobs/lib/routes/tasks/putTask.ts]: 200 seems more correct

File: packages/runner/lib/server.ts [hunk 10]
```diff
@@ -92,18 +96,31 @@ function startProcedure() {
             // executing in the background and returning immediately
             // sending the result to the jobs service when done
             setImmediate(async () => {
+                const heartbeat = setInterval(async () => {
+                    await httpFetch({
+                        method: 'POST',
+                        url: `${jobsServiceUrl}/tasks/${taskId}/heartbeat`
+                    });
+                }, heartbeatIntervalMs);
```
Comment (bodinsamuel) [inline packages/runner/lib/server.ts:104]: On v1 I think it's fine, but could probably be safer on a different thread to avoid missing a beat

File: packages/orchestrator/lib/abort.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import type { Scheduler, Task } from '@nangohq/scheduler';
+import { validateTask } from './clients/validate.js';
+import { Ok, stringifyError } from '@nangohq/utils';
+import type { Result } from '@nangohq/utils';
+import { logger } from './utils.js';
+
+export async function scheduleAbortTask({ scheduler, task }: { scheduler: Scheduler; task: Task }): Promise<Result<Task | undefined>> {
+    const aborted = validateTask(task);
+    if (aborted.isOk()) {
```
Comment (bodinsamuel) [inline packages/orchestrator/lib/abort.ts]: You are not rejecting if the validation doesn't pass?

File: packages/jobs/lib/routes/postIdle.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            .object({ runnerId: z.string().min(1), idleTimeMs: z.number().positive() })
+            .strict()
+            .parse(data)
+});
+
+const handler = async (req: EndpointRequest<PostIdle>, res: EndpointResponse<PostIdle>) => {
+    try {
+        logger.info(`[runner ${req.body.runnerId}]: idle for ${req.body.idleTimeMs}ms. Suspending...`);
+        await suspendRunner(req.body.runnerId);
+        return res.status(201).json({ status: 'ok' });
```
Comment (bodinsamuel) [inline packages/jobs/lib/routes/postIdle.ts]: Also 200?

File: packages/jobs/lib/execution/action.ts [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+            activityLogId: task.activityLogId,
+            runTime: 0,
+            error,
+            environment: { id: task.connection.environment_id, name: environment?.name || 'unknown' },
+            ...(account?.id && account?.name ? { team: { id: account.id, name: account.name } } : {})
+        });
+        return Err(error);
+    }
+}
+export async function handleActionOutput({ nangoProps }: { nangoProps: NangoProps }): Promise<void> {
```
Comment (bodinsamuel) [inline packages/jobs/lib/execution/action.ts]: I would have named handle*Success to be consistent with handle*Error

**Review by bodinsamuel — COMMENTED** [review]
Looks good, tested locally to sync run/cancel, action run.
- Some integration test on the API would be great
- I had multiple times `hasn't started after 5000ms` maybe the `waitForRunner` is not reactive enough

Few questions, but seems like it's mostly good

File: packages/cli/lib/services/dryrun.service.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/cli/lib/services/dryrun.service.ts:505]: no, they can't really but might a nice addition in the future

File: packages/jobs/lib/routes/tasks/putTask.ts (hunk 3 above)
Comment (TBonnin) [reply packages/jobs/lib/routes/tasks/putTask.ts:88]: good point. We don't have to recreate it every time. will change

File: packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts (hunk 6 above)
Comment (TBonnin) [reply packages/jobs/lib/routes/tasks/taskId/postHeartbeat.ts:29]: sorry. what do you mean?

File: packages/jobs/lib/routes/tasks/putTask.ts (hunk 9 above)
Comment (TBonnin) [reply packages/jobs/lib/routes/tasks/putTask.ts]: actually a 204 no content might even be better

File: packages/runner/lib/server.ts (hunk 10 above)
Comment (TBonnin) [reply packages/runner/lib/server.ts:104]: indeed. However the heartbeat timeout is much bigger than the hearbeat interval so we can miss a lot of them without consequences

File: packages/orchestrator/lib/abort.ts (hunk 11 above)
Comment (TBonnin) [reply packages/orchestrator/lib/abort.ts]: you are right. don't remember why i did that but that doesn't seem correct. will change to return an error if it doesn't validate

File: packages/jobs/lib/execution/action.ts (hunk 13 above)
Comment (TBonnin) [reply packages/jobs/lib/execution/action.ts]: good idea

File: packages/shared/lib/sdk/sync.integration.test.ts (hunk 2 above)
Comment (TBonnin) [reply packages/shared/lib/sdk/sync.integration.test.ts]: let me see. if this is feasible right now

Comment (TBonnin) [reply packages/shared/lib/sdk/sync.integration.test.ts]: done

**Comment by bodinsamuel** [comment]
Ping me when you want a last review 👌🏻

**Review by bodinsamuel — APPROVED** [review]
🎸
